cclib/parser/gamessdatparser.py

```diff
@@ -1,246 +1,314 @@
-        # EXPONENTS  1.6885540E-01
\ No newline at end of file
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2023, the cclib development team
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+"""Parser for GAMESS(US) .dat output files"""
+
+
+import re
+import numpy
+
+from cclib.parser import logfileparser
+from cclib.parser import utils
+
+from cclib.parser.utils import PeriodicTable
+
+class GAMESSDAT(logfileparser.Logfile):
+    """A GAMESS .dat log file"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(logname="GAMESSDAT", *args, **kwargs)
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return f"GAMESS .dat log file {self.filename}"
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return f'GAMESS .dat ("{self.filename}")'
+    
+    def normalisesym(self, label):
+        """Normalise the symmetries used by GAMESS .dat."""
+
+        pass
+
+    def before_parsing(self):
+        self.pt = PeriodicTable()
+        # To change: declared only for passing unit tests
+
+    def after_parsing(self):
+        if hasattr(self, "atomcoords"):
+            len_coords = len(self.atomcoords)
+            self.atomcoords = numpy.reshape(self.atomcoords, (1, len_coords, 3))
+        
+        if hasattr(self, "mocoeffs"):
+            self.mocoeffs = numpy.reshape(self.mocoeffs, (1, self.nmo, self.nbasis))
+
+
+    def extract(self, inputfile, line):
+        """Extract information from the file object inputfile."""
+        
+        # Extract element and its properties
+
+        # $DATA  
+        # water                                                                           
+        # C1       0
+        # O           8.0      0.0000000000      0.0000000000      0.0000000000
+        #    STO     3
+                
+        # H           1.0      0.9900000000      0.0000000000      0.0000000000
+        #    STO     3
+                
+        # H           1.0     -0.2728810000      0.9516490000      0.0000000000
+        #    STO     3
+                
+        #  $END   
+
+        # Extract molecule name
+
+        if line[1:6] == "$DATA":
+
+            line = next(inputfile)
+            name = line.split('\n')[0].strip()
+            self.metadata["name"] = name
+
+            # Extract atomic information
+
+            self.atommasses = []
+            self.atomcoords = []
+
+            line = next(inputfile)
+
+            while line.strip() != "$END":
+                parts = line.split()
+                if len(parts) > 2:
+
+                    symbol      = parts[0]
+                    mass        = float(parts[1])
+                    coordinates = [float(coord) for coord in parts[2:]]
+
+                    assert len(coordinates) == 3
+
+                    self.atommasses.append(mass)

```

Comment:
Might not be the best place to suggest it, but it would actually be pretty cool if the `self.extend_attribute(...)` [function](https://github.com/cclib/cclib/blob/2b947c1d0d931d11d1b6fcf5b08eea5999a418a2/cclib/parser/logfileparser.py#L479C9-L479C15) had an additional optional parameter for the index (or have a new function that does it). I could use it instead of having to write `self.mocoeffs[-1].extend(mocoeff)`. 

I could even give it a shot and try to implement it myself if that sounds like a good idea.

Reply:
I think this is a good idea in one form or another.  Can you make a new issue?  There are lots of other things that these `self.*_attribute` could use, but we haven't kept good track of them.